source/FreeRTOS_Routing.c

```diff
@@ -1545,3 +1551,250 @@ struct xIPv6_Couple
     }
 /*-----------------------------------------------------------*/
 #endif /* ( ( ipconfigHAS_PRINTF != 0 ) || ( ipconfigHAS_DEBUG_PRINTF != 0 ) ) */
+
+/**
+ * @brief Check whether a packet needs resolution if it is on local subnet. If required send a request.
+ *
+ * @param[in] pxNetworkBuffer The network buffer with the packet to be checked.
+ *
+ * @return pdTRUE if the packet needs resolution, pdFALSE otherwise.
+ */
+BaseType_t xCheckRequiresResolution( const NetworkBufferDescriptor_t * const pxNetworkBuffer )
+{
+    BaseType_t xNeedsResolution = pdFALSE;
+
+    switch( usFrameType( pxNetworkBuffer ) )
+    {
+        case ipIPv4_FRAME_TYPE:
+            #if ipconfigIS_ENABLED( ipconfigUSE_IPv4 )
+                xNeedsResolution = xCheckIPv4RequiresResolution( pxNetworkBuffer );
+            #else
+                FreeRTOS_debug_printf( ( "xCheckRequiresResolution: Unsupported IPv4 Packet\n" ) );
+            #endif
+            break;
+
+        case ipIPv6_FRAME_TYPE:
+            #if ipconfigIS_ENABLED( ipconfigUSE_IPv6 )
+                xNeedsResolution = xCheckIPv6RequiresResolution( pxNetworkBuffer );
+            #else
+                FreeRTOS_debug_printf( ( "xCheckRequiresResolution: Unsupported IPv6 Packet\n" ) );
+            #endif
+            break;
+
+        default:
+            /* Shouldn't reach here */
+            FreeRTOS_debug_printf( ( "xCheckRequiresResolution: Invalid Packet Type\n" ) );
+            configASSERT( pdFALSE );
+            /* MISRA 16.4 Compliance */
+            break;
+    }
+
+    return xNeedsResolution;
+}
+/*-----------------------------------------------------------*/
+
+#if ipconfigIS_ENABLED( ipconfigUSE_IPv6 )
+
+/**
+ * @brief Get the first end point of the type (IPv4/IPv6) from the list
+ *        the list of end points.
+ *
+ * @param[in] xIPType IP type (ipTYPE_IPv6/ipTYPE_IPv4)
+ * @param[in] xIPAddrType Specifies required endpoint type.
+ *
+ * @returns Pointer to the first end point of the given IP type from the
+ *          list of end points.
+ */
+    static NetworkEndPoint_t * prvGetEndpointFromType( BaseType_t xIPType,
+                                                       IPv6_Type_t xIPAddrType )
+    {
+        NetworkEndPoint_t * pxResult = NULL;
+        NetworkEndPoint_t * pxEndPoint;
+
+        for( pxEndPoint = FreeRTOS_FirstEndPoint( NULL );
+             pxEndPoint != NULL;
+             pxEndPoint = FreeRTOS_NextEndPoint( NULL, pxEndPoint ) )
+        {
+            if( xIPType == ( BaseType_t ) ipTYPE_IPv6 )
+            {
+                #if ipconfigIS_ENABLED( ipconfigUSE_IPv6 )
+                    if( pxEndPoint->bits.bIPv6 != 0U )
+                    {
+                        const IPv6_Type_t eEndpointType = xIPv6_GetIPType( &( pxEndPoint->ipv6_settings.xIPAddress ) );
+                        if( eEndpointType == xIPAddrType )
```

Comment:
Why do we need endpoints that perfectly match the address type of destination or source address? For loopback addresses, this means that we need an endpoint with loopback address as its IP.

Reply:
This was intended to replace functions like pxFindLocalEndpoint, you are correct it's not valid here and is something I have removed (yet to push) to finish later to keep this PR simpler.